lib/sinatra/trusted_hosts.rb

```diff
@@ -6,17 +7,59 @@ def authorize_with_trusted_hosts
-        # When :trusted_hosts is given, check the reverse DNS entry of the remote IP and ensure it's listed
+        # When :trusted_hosts is given, we check the client against the list
+        # HTTPS: test the certificate CN
+        # HTTP: test the reverse DNS entry of the remote IP
         if Proxy::SETTINGS.trusted_hosts
-          begin
-            remote_fqdn = Resolv.new.getname(request.env["REMOTE_ADDR"])
-          rescue Resolv::ResolvError => e
-            log_halt 403, "Unable to resolve hostname for connecting client - #{request.env["REMOTE_ADDR"]}. If it's to be a trusted host, ensure it has a reverse DNS entry."  +
-            "\n\n" + "#{e.message}"
+
+          remote_ip = request.env["REMOTE_ADDR"]
+
+          logger.debug "verifying remote client #{remote_ip} against trusted_hosts #{Proxy::SETTINGS.trusted_hosts.to_s}"
+
+          if ['yes', 'on', '1'].include? request.env['HTTPS'].to_s
+
+            certificate_raw = request.env['SSL_CLIENT_CERT'].to_s
+            if certificate_raw.empty?
+              log_halt 403, "Could not get client cert from environment!"
+            end
+
+            certificate = OpenSSL::X509::Certificate.new certificate_raw
+            if certificate.subject && certificate.subject.to_s =~ /CN=([^\s\/,]+)/i
+              remote_fqdn = $1
+              logger.debug "found remote CN #{remote_fqdn}"
+            else
+              log_halt 403, "Could not get client CN from the certificate!"
+            end
+
+          else
+
+            begin
+              remote_fqdn = Resolv.new.getname(remote_ip)
+              logger.debug "Remote reverse lookup is #{remote_fqdn}"
+            rescue Resolv::ResolvError => e
+              log_halt 403, "Unable to resolve hostname for connecting client - #{remote_ip}. If it's to be a trusted host, ensure it has a reverse DNS entry." +
+              "\n\n" + "#{e.message}"
+            end
+
+            begin
+              forward_lookup = Resolv.new.getaddresses(remote_fqdn)
+              logger.debug "Remote forward lookup is #{forward_lookup.to_s}"
+            rescue Resolv::ResolvError => e
+              log_halt 403, "Unable to forward verify the reverse DNS entry of the connected client - #{remote_fqdn} (#{remote_ip}). If it's to be a trusted host, ensure it has a reverse DNS and a matching forward entry." +
+              "\n\n" + "#{e.message}"
+            end
+
+            if !forward_lookup.include?(remote_ip)
+              log_halt 403, "Untrusted client has no matching forward DNS lookup - #{remote_fqdn} (#{remote_ip}). If it's to be a trusted host, ensure it has a reverse DNS and a matching forward entry." +
+              "\n\n" + "#{e.message}"
+            end
+
           end
+
           if !Proxy::SETTINGS.trusted_hosts.include?(remote_fqdn.downcase)
             log_halt 403, "Untrusted client #{remote_fqdn.downcase} attempted to access #{request.path_info}. Check :trusted_hosts: in settings.yml"
           end
+
```

Comment:
Why do we just throw the `remote_fqdn` out on line 37. I understand that client can send any certificate he wants with arbitrary CN, but at least for correct queries both CA name and reverse lookup should both match.